### PR TITLE
improve assumereadtype

### LINKED ISSUE #3300: Further improve mongo's AssumeReadType
`AssumeReadType` does not recognize all shape preserving constructions yet - see https://github.com/quasar-analytics/quasar/pull/3240#issuecomment-350229794 so there are still guards left that can be elided. Investigate whether this can be improved further.

### PR DESCRIPTION
**Pull request by rintcius**
Create a new typeclass `ShapePreserving` and use that to further improve `assumeReadType`.

Fixes #3300

### CODE REVIEW

**Comment by rintcius** [comment]
@wemrysi like you mentioned in qz-3300 there's a more reliable interim way than proceeding with the current mechanism, but given that this is there already (and looking pretty good now), we may be able to skip that interim way

**Comment by jsantos17** [comment]
I think the typechecks for this query can be simplified too:

```
select codes[*] as codes , first_name from `/mngo/jps/patients` where state = "CO"
```

QScript:

```
QScript Mongo (Prefer Projection):
LeftShift
├─ ShiftedRead(/jps/patients, ExcludeId)
├─ Struct
│  ╰─ Guard
│     ├─ ProjectKey
│     │  ├─ Cond
│     │  │  ├─ Eq
│     │  │  │  ├─ ProjectKey
│     │  │  │  │  ├─ ○
│     │  │  │  │  ╰─ Constant(Str(state))
│     │  │  │  ╰─ Constant(Str(CO))
│     │  │  ├─ ○
│     │  │  ╰─ Undefined
│     │  ╰─ Constant(Str(codes))
│     ├─ Type(FlexArr(0, None, Top))
│     ├─ ProjectKey
│     │  ├─ Cond
│     │  │  ├─ Eq
│     │  │  │  ├─ ProjectKey
│     │  │  │  │  ├─ ○
│     │  │  │  │  ╰─ Constant(Str(state))
│     │  │  │  ╰─ Constant(Str(CO))
│     │  │  ├─ ○
│     │  │  ╰─ Undefined
│     │  ╰─ Constant(Str(codes))
│     ╰─ Undefined
├─ IdStatus
│  ╰─ IdStatus(ExcludeId)
├─ ShiftType
│  ╰─ ShiftType(Array)
╰─ Repair
   ╰─ ConcatMaps
      ├─ MakeMap
      │  ├─ Constant(Str(codes))
      │  ╰─ RightSide
      ╰─ MakeMap
         ├─ Constant(Str(first_name))
         ╰─ Guard
            ├─ LeftSide
            ├─ Type(Obj(Map[], Some(Top)))
            ├─ ProjectKey
            │  ├─ Cond
            │  │  ├─ Eq
            │  │  │  ├─ ProjectKey
            │  │  │  │  ├─ LeftSide
            │  │  │  │  ╰─ Constant(Str(state))
            │  │  │  ╰─ Constant(Str(CO))
            │  │  ├─ LeftSide
            │  │  ╰─ Undefined
            │  ╰─ Constant(Str(first_name))
            ╰─ Undefined
```

I think the typecheck on the `LeftSide` of the repair is known statically.

File: connector/src/main/scala/quasar/qscript/analysis/ShapePreserving.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import simulacrum._
+
+/** Determines if `IN[_]` preserves the shape of its underlying `ShiftedRead`.
+  * If it does preserve that shape then it returns the `IdStatus` of that
+  * `ShiftedRead`. If it doesn't then `none` is returned.
+  * This can easily be extended to also include `Read`, possibly with extra info
+  * to indicate whether the underlying read was `Read` or `ShiftedRead`.
+  */
+@typeclass
+trait ShapePreserving[IN[_]] {
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/analysis/ShapePreserving.scala]: The name here is a bit too general, I think, maybe something like `PreserveRead` or similar.

Comment (rintcius) [reply connector/src/main/scala/quasar/qscript/analysis/ShapePreserving.scala]: Ah yeah, `PreserveRead` is better 👍

**Pull request merged**